src/sentry/snuba/query_subscription_consumer.py

```diff
@@ -58,42 +72,59 @@ def inner(func: TQuerySubscriptionCallable) -> TQuerySubscriptionCallable:
         try:
-            jsonschema.validate(wrapper, SUBSCRIPTION_WRAPPER_SCHEMA)
-        except jsonschema.ValidationError:
-            metrics.incr("snuba_query_subscriber.message_wrapper_invalid")
-            raise InvalidSchemaError("Message wrapper does not match schema")
+            jsoncodec = JsonCodec(get_schema(topic)["schema"])
+            jsoncodec.validate(wrapper)
+        except ValidationError:
+            old_version = True
+            metrics.incr("snuba_query_subscriber.message_wrapper.old_validation")
+            try:
+                jsonschema.validate(wrapper, SUBSCRIPTION_WRAPPER_SCHEMA)
+            except jsonschema.ValidationError:
+                metrics.incr("snuba_query_subscriber.message_wrapper_invalid")
+                raise InvalidSchemaError("Message wrapper does not match schema")
 
     schema_version: int = wrapper["version"]
-    if schema_version not in SUBSCRIPTION_PAYLOAD_VERSIONS:
-        metrics.incr("snuba_query_subscriber.message_wrapper_invalid_version")
-        raise InvalidMessageError("Version specified in wrapper has no schema")
+    payload: PayloadV3 = wrapper["payload"]
+
+    if old_version:
+        if schema_version not in SUBSCRIPTION_PAYLOAD_VERSIONS:
+            metrics.incr("snuba_query_subscriber.message_wrapper_invalid_version")
+            raise InvalidMessageError("Version specified in wrapper has no schema")
+
+        with metrics.timer("snuba_query_subscriber.parse_message_value.json_validate_payload"):
+            try:
+                jsonschema.validate(payload, SUBSCRIPTION_PAYLOAD_VERSIONS[schema_version])
+            except jsonschema.ValidationError:
+                metrics.incr("snuba_query_subscriber.message_payload_invalid")
+                raise InvalidSchemaError("Message payload does not match schema")
+
+    else:
+        if schema_version != SUPPORTED_SCHEMA_VERSION:
+            metrics.incr("snuba_query_subscriber.message_wrapper_invalid_version")
+            raise InvalidMessageError("Version specified in wrapper has no schema")
 
-    payload: Dict[str, Any] = wrapper["payload"]
-    with metrics.timer("snuba_query_subscriber.parse_message_value.json_validate_payload"):
-        try:
-            jsonschema.validate(payload, SUBSCRIPTION_PAYLOAD_VERSIONS[schema_version])
-        except jsonschema.ValidationError:
-            metrics.incr("snuba_query_subscriber.message_payload_invalid")
-            raise InvalidSchemaError("Message payload does not match schema")
     # XXX: Since we just return the raw dict here, when the payload changes it'll
     # break things. This should convert the payload into a class rather than passing
     # the dict around, but until we get time to refactor we can keep things working
     # here.
-    payload.setdefault("values", payload.get("result"))
+    payload.setdefault("values", payload.get("result"))  # type: ignore
```

Comment:
Why do we have to do `type: ignore` here and the line below?

Reply:
Maybe that can be changed - typing was yelling that we were adding properties that aren't in the defined types generated by the schema, but at this point in the code we don't care - we just want to make sure it matches the schema when it comes in. Alternately I could make a copy of the object to avoid ignoring it?

Reply:
Not sure I understand why the `result` is copied over into the `values` key anyway